                                                                       TEMUULEN Bulgan - 2022427

### CCT College Dublin Continuous Assessment No.2
# AN ANALYSIS OF INDIAN FARMERS' PROTEST TWEETS.

**Brief Introduction**

1. For my second continuous assessment, I choose CSV format data of Indian Farmer's Protest Tweets. This file contains over 1 million English language tweets tweeted between November 1st, 2020 and november 21st, 2021 with the hashtag <#FarmersProtest>. It is downloaded from the Kaggle website with the CCO:Public Domain license.
(https://kaggle.com/datasets/prathamsharma123/farmers-protest-tweets-dataset-csv)

2. I divided my project into 3 primary sections (Every step in data processing and analysis is fully discussed on each subsection of these primary sections.):
    1. big data storage and processing
    2. comparative analysis of databases
    3. sentiment analysis and forecast

3. I used Git for daily code tracking and GitHub for archiving, monitoring and sharing. To view the whole project on GitHub, click on the follwig link 

In [32]:
# First of all, importing libraries that I used in this project.
import pandas as pd
import os
from pyspark.sql import SparkSession
from pyspark.sql.functions import *

## 1. Big Data Storage and Processing

Since the size of the data I choose is 1.8 GB it would be difficult for me to upload and process it on my VM, I decided to do some preprocessing in Python's Pandas library.

In [3]:
# Checking the file size of the file.
file_size = os.path.getsize('/home/hduser/Desktop/ca/tweets.csv')/(1024*1024*1024)

print('The size of the tweet file is ', round(file_size, 2), 'GB.')

The size of the tweet file is  1.69 GB.


### 1.1. *Dataset prepocessing in Python.*

Because the DataSet size is 1.69 GB it will be difficult to process it in my VM environment, that's why I'm going to run the file on the Jupyter Notebook and check the information content. 

In [5]:
# Creating a DataFrame.
df_preprocess = pd.read_csv('/home/hduser/Desktop/ca/tweets.csv')

# Checking the DataFrame.
df_preprocess

tweetUrl   
0        https://twitter.com/ShashiRajbhar6/status/1376...  \
1        https://twitter.com/kaursuk06272818/status/137...   
2        https://twitter.com/kaursuk06272818/status/137...   
3        https://twitter.com/SukhdevSingh_/status/13767...   
4        https://twitter.com/Davidmu66668113/status/137...   
...                                                    ...   
1084447  https://twitter.com/drkumarpmch2000/status/144...   
1084448  https://twitter.com/DigitalKisanBot/status/144...   
1084449  https://twitter.com/sandhumannu0476/status/144...   
1084450  https://twitter.com/mani262002/status/14433695...   
1084451  https://twitter.com/sandhumannu0476/status/144...   

                              date   
0        2021-03-30 03:33:46+00:00  \
1        2021-03-30 03:33:23+00:00   
2        2021-03-30 03:31:00+00:00   
3        2021-03-30 03:30:45+00:00   
4        2021-03-30 03:30:30+00:00   
...                            ...   
1084447  2021-09-30 00:48:05+00:00   
1084448  2021-09-30 00:28:06+00:00   
1084449  2021-09-30 00:22:20+00:00   
1084450  2021-09-30 00:18:06+00:00   
1084451  2021-09-30 00:13:02+00:00   

                                           renderedContent       tweetId   
0                             Support 👇\n\n#FarmersProtest  1.376739e+18  \
1        Supporting farmers means supporting our countr...  1.376739e+18   
2        Support farmers if you are related to food #St...  1.376739e+18   
3        #StopHateAgainstFarmers support #FarmersProtes...  1.376739e+18   
4        You hate farmers I hate you, \nif you love the...  1.376739e+18   
...                                                    ...           ...   
1084447            Good morning friends 🌹\n#FarmersProtest  1.443377e+18   
1084448  There is no compromise when it comes to injust...  1.443372e+18   
1084449  @KhalsaAid_AU @Khalsa_Aid ਮਾਨਸ ਕੀ ਜਾਤ ਸਭੈ ਏਕੈ ...  1.443371e+18   
1084450  #FarmersProtest_StandsStrong \n#FarmersProtest...  1.443370e+18   
1084451  Good morning friends... \n#FarmersProtest http...  1.443368e+18   

               userId  replyCount  retweetCount  likeCount  quoteCount   
0        1.015970e+18           0             0          0           0  \
1        1.332937e+18           0             0          0           0   
2        1.332937e+18           0             0          0           0   
3        1.308357e+18           0             1          3           0   
4        1.357312e+18           0             0          1           0   
...               ...         ...           ...        ...         ...   
1084447  3.035365e+09           1             1          1           0   
1084448  1.309303e+18           0             0          1           0   
1084449  1.377187e+18           0             0          0           0   
1084450  2.350986e+09           0             0          0           0   
1084451  1.377187e+18           0             0          0           0   

                                                    source   
0        <a href="http://twitter.com/download/android" ...  \
1        <a href="http://twitter.com/download/android" ...   
2        <a href="http://twitter.com/download/android" ...   
3        <a href="http://twitter.com/download/android" ...   
4        <a href="http://twitter.com/download/android" ...   
...                                                    ...   
1084447  <a href="http://twitter.com/download/android" ...   
1084448  <a href="https://help.twitter.com/en/using-twi...   
1084449  <a href="http://twitter.com/download/android" ...   
1084450  <a href="https://mobile.twitter.com" rel="nofo...   
1084451  <a href="http://twitter.com/download/android" ...   

                                                     media  retweetedTweet   
0                                                      NaN             NaN  \
1        [{'previewUrl': 'https://pbs.twimg.com/media/E...             NaN   
2        [{'previewUrl': 'https://pbs.twimg.com/media/E...         

In [6]:
# Examaning DataFrame.
df_preprocess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1084452 entries, 0 to 1084451
Data columns (total 14 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   tweetUrl         1084452 non-null  object 
 1   date             1084452 non-null  object 
 2   renderedContent  1084452 non-null  object 
 3   tweetId          1084452 non-null  float64
 4   userId           1084452 non-null  float64
 5   replyCount       1084452 non-null  int64  
 6   retweetCount     1084452 non-null  int64  
 7   likeCount        1084452 non-null  int64  
 8   quoteCount       1084452 non-null  int64  
 9   source           1072318 non-null  object 
 10  media            319141 non-null   object 
 11  retweetedTweet   0 non-null        float64
 12  quotedTweet      286988 non-null   object 
 13  mentionedUsers   412206 non-null   object 
dtypes: float64(3), int64(4), object(7)
memory usage: 115.8+ MB


This dataset contains 8 distinct categories of information that I will use for further analysis. Rest is not important, it will use only memory and space of my VM, that's why i'm going to remove them.

In [7]:
# Deleting columns from the DataFrame
df_preprocess = df_preprocess.drop(labels=['tweetUrl', 'tweetId', 'source', 'media', 'retweetedTweet', 'quotedTweet', \
                                           'mentionedUsers'], axis=1)

# Checking the changes.
df_preprocess.head(2)

date   
0  2021-03-30 03:33:46+00:00  \
1  2021-03-30 03:33:23+00:00   

                                     renderedContent       tweetId   
0                       Support 👇\n\n#FarmersProtest  1.376739e+18  \
1  Supporting farmers means supporting our countr...  1.376739e+18   

         userId  replyCount  retweetCount  likeCount  quoteCount  
0  1.015970e+18           0             0          0           0  
1  1.332937e+18           0             0          0           0

I'm saving the file now as a 'new_tweets.csv' and uploadig it into the Hadoop HDFS where I'll do some extra processing using Apache Spark tools.

In [8]:
# Creating a new file of tweets on my VM with the utf-8 Encoding.
df_preprocess.to_csv('/home/hduser/Desktop/ca/new_tweets.csv', encoding='utf-8', index=False)

In [9]:
# Checking the file size.
file_size = os.path.getsize('/home/hduser/Desktop/ca/new_tweets.csv')/(1024*1024)
print('The size of the tweet file is ', round(file_size, 2), 'MB.')

The size of the tweet file is  265.58 MB.


In [48]:
# Checking the DataFrame.
df_preprocessed = pd.read_csv('/home/hduser/Desktop/ca/new_tweets.csv')
df_preprocessed.head()

date   
0  2021-03-30 03:33:46+00:00  \
1  2021-03-30 03:33:23+00:00   
2  2021-03-30 03:31:00+00:00   
3  2021-03-30 03:30:45+00:00   
4  2021-03-30 03:30:30+00:00   

                                     renderedContent       tweetId   
0                       Support 👇\n\n#FarmersProtest  1.376739e+18  \
1  Supporting farmers means supporting our countr...  1.376739e+18   
2  Support farmers if you are related to food #St...  1.376739e+18   
3  #StopHateAgainstFarmers support #FarmersProtes...  1.376739e+18   
4  You hate farmers I hate you, \nif you love the...  1.376739e+18   

         userId  replyCount  retweetCount  likeCount  quoteCount  
0  1.015970e+18         0.0           0.0        0.0         0.0  
1  1.332937e+18         0.0           0.0        0.0         0.0  
2  1.332937e+18         0.0           0.0        0.0         0.0  
3  1.308357e+18         0.0           1.0        3.0         0.0  
4  1.357312e+18         0.0           0.0        1.0         0.0

In [49]:
def edit_val(val):
    for x in range(len(val)):
        if val[x] == ' ':
            return val[0:x]
            
df_preprocessed['date'] = df_preprocessed['date'].apply(edit_val)


In [71]:
df_preprocessed.tail()

date                                    renderedContent   
1084453  2021-09-30            Good morning friends 🌹\n#FarmersProtest  \
1084454  2021-09-30  There is no compromise when it comes to injust...   
1084455  2021-09-30  @KhalsaAid_AU @Khalsa_Aid ਮਾਨਸ ਕੀ ਜਾਤ ਸਭੈ ਏਕੈ ...   
1084456  2021-09-30  #FarmersProtest_StandsStrong \n#FarmersProtest...   
1084457  2021-09-30  Good morning friends... \n#FarmersProtest http...   

              tweetId        userId  replyCount  retweetCount  likeCount   
1084453  1.443377e+18  3.035365e+09         1.0           1.0        1.0  \
1084454  1.443372e+18  1.309303e+18         0.0           0.0        1.0   
1084455  1.443371e+18  1.377187e+18         0.0           0.0        0.0   
1084456  1.443370e+18  2.350986e+09         0.0           0.0        0.0   
1084457  1.443368e+18  1.377187e+18         0.0           0.0        0.0   

         quoteCount  
1084453         0.0  
1084454         0.0  
1084455         0.0  
1084456         0.0  
1084457         0.0

Now the file size is reduced to 272.8 and we still have importand information of the tweets for further processing.

### 1.2 *Dataset processing in Pyspark.*

I've copied the 'new_tweet.csv' file from the local file system to HDFS's new created directory 'ca2'. And now I would like to do some processing job to the file using Pyspark library.

In [63]:
# Starting a new Sparksession for data import from HDFS.
spark = SparkSession.builder \
        .appName('HDFS Data Import') \
        .getOrCreate()

In [64]:
# Readig the file.
df_spark = spark.read.option('header', 'true') \
                        .csv('/ca2/new_tweets.csv')
#            .option('inferSchema', 'true') \
#            .option('enforeSchema', 'true') \
           

In [70]:
df_spark.show(100)

+--------------------+--------------------+---------+-------+---------+-----+------+
|                date|               tweet|     user|replyed|retweeted|liked|quoted|
+--------------------+--------------------+---------+-------+---------+-----+------+
|2021-03-30 03:33:...|          Support 👇|     null|   null|     null| null|  null|
|    #FarmersProtest"|1.376739399593910...|        0|      0|        0|    0|  null|
|2021-03-30 03:33:...|Supporting farmer...|     null|      0|        0|    0|     0|
|2021-03-30 03:31:...|Support farmers i...|     null|      0|        0|    0|     0|
|2021-03-30 03:30:...|#StopHateAgainstF...|     null|      0|        1|    3|     0|
|2021-03-30 03:30:...|You hate farmers ...|     null|   null|     null| null|  null|
|if you love them ...|                null|     null|   null|     null| null|  null|
|#StopHateAgainstF...|                null|     null|   null|     null| null|  null|
|    #FarmersProtest"|1.376738579171344...|        0|      0|     

In [68]:
#
df_spark.printSchema()

root
 |-- date: string (nullable = true)
 |-- tweet: string (nullable = true)
 |-- user: integer (nullable = true)
 |-- replyed: integer (nullable = true)
 |-- retweeted: integer (nullable = true)
 |-- liked: integer (nullable = true)
 |-- quoted: integer (nullable = true)



In [67]:
# Renaming the columns.
df_spark = df_spark.select(col('date').alias('date'),
                           col('renderedContent').alias('tweet'),
                           col('userId').alias('user'),
                           col('replyCount').alias('replyed'),
                           col('retweetCount').alias('retweeted'),
                           col('likeCount').alias('liked'),
                           col('quoteCount').alias('quoted'))

In [66]:

#df_spark = df_spark.withColumn('date',to_date(col('date').cast('yyyy-MM-dd')))
df_spark = df_spark.withColumn('userId',col('userId').cast('integer'))
df_spark = df_spark.withColumn('replyCount',col('replyCount').cast('integer'))
df_spark = df_spark.withColumn('retweetCount',col('retweetCount').cast('integer'))
df_spark = df_spark.withColumn('likeCount',col('likeCount').cast('integer'))
df_spark = df_spark.withColumn('quoteCount',col('quoteCount').cast('integer'))

In [11]:
print(type(df_spark.count()))

<class 'int'>


In [36]:
spark.sql('select * from tweet').show(4)

+--------------------+--------------------+--------------------+------+----------+------------+---------+----------+
|                date|     renderedContent|             tweetId|userId|replyCount|retweetCount|likeCount|quoteCount|
+--------------------+--------------------+--------------------+------+----------+------------+---------+----------+
|2021-03-30 03:33:...|          Support 👇|                null|  null|      null|        null|     null|      null|
|    #FarmersProtest"|1.376739399593910...|1.015969769760096...|     0|         0|           0|        0|      null|
|2021-03-30 03:33:...|Supporting farmer...|1.376739306287427...|  null|         0|           0|        0|         0|
|2021-03-30 03:31:...|Support farmers i...|1.376738704128020...|  null|         0|           0|        0|         0|
+--------------------+--------------------+--------------------+------+----------+------------+---------+----------+
only showing top 4 rows

